pkg/controller/operators/catalog/operator.go

```diff
@@ -1125,10 +1145,46 @@ func (o *Operator) validateExistingCRs(gvr schema.GroupVersionResource, newCRD *
 			return fmt.Errorf("error validating custom resource against new schema %#v: %s", newCRD.Spec.Validation, err)
 		}
 	}
-
 	return nil
 }
 
+// Attempt to remove stored versions that have been deprecated before allowing
+// those versions to be removed from the new CRD.
+// The functiona may not always succeed as storedVersions requires at least one
+// version. If there is only stored version, it won't be removed until a new
+// stored version is added.
+func removeDeprecatedStoredVersions(oldCRD *v1beta1ext.CustomResourceDefinition, newCRD *v1beta1ext.CustomResourceDefinition) *v1beta1ext.CustomResourceDefinition {
+	// StoredVersions requires to have at least one version.
+	if len(oldCRD.Status.StoredVersions) <= 1 {
+		return nil
+	}
+
+	storedVersionsMap := map[string]int{}
+	for i, v := range oldCRD.Status.StoredVersions {
+		storedVersionsMap[v] = i
+	}
+
+	newCRDVersions := getCRDVersionsMap(newCRD)
+
+	for _, version := range oldCRD.Status.StoredVersions {
+		_, ok := newCRDVersions[version]
+		if !ok {
+			delete(storedVersionsMap, version)
```

Comment:
Instead of doing this delete, you could combine with the above for loop and just insert when the old version is also present in the newCRDVersions map.